<a href="https://colab.research.google.com/github/dorvalneto/IA_experimental/blob/main/Notebook_An%C3%A1lise_de_dados_com_Python_utilizando_o_ChatGPT_como_assistente.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

![Alt text: Logo da Zoop](https://github.com/alura-cursos/python-analise-chatgpt-assistente/blob/main/Logos%20e%20paleta%20-%20ZOOP/Logo%20fundo%20colorido%20(2).png?raw=true)

# Aula 1 - Acesso aos dados

In [1]:
url = 'https://github.com/alura-cursos/python-analise-chatgpt-assistente/raw/main/Dados/dados_vendas.json'

### Prompt:

Vamos atuar como cientistas de dados em uma empresa de supermercados.

Recebemos dados no formato de arquivo JSON, estruturados em listas e dicionários. O arquivo está disponibilizado em uma URL do Github.

Precisamos analisar esses dados para ajudar nas decisões da empresa. Programamos em Python e conhecemos as bibliotecas de análise de dados Numpy e Pandas. Portanto, priorize o uso dessas bibliotecas.

Por favor, nos informe como podemos fazer a leitura desses dados no formato JSON e como podemos visualizá-los em formato de tabela no Python.

In [2]:
import requests
import pandas as pd
import numpy as np


In [3]:
response = requests.get(url)
data = response.json()

In [5]:
df = pd.DataFrame(data)
df.head()  # Exibe as primeiras 5 linhas do DataFrame


item_identificador loja_identificador  vendas_totais  \
0              FDB08             OUT018      176503.58   
1              DRQ35             OUT049      185758.20   
2              FDD14             OUT018      165983.94   
3              FDY37             OUT045      314923.40   
4              FDY59             OUT018       64782.34   

                                                item  \
0  {'item_peso': 6.055, 'item_conteudo_gordura': ...   
1  {'item_peso': 9.3, 'item_conteudo_gordura': 'B...   
2  {'item_peso': 20.7, 'item_conteudo_gordura': '...   
3  {'item_peso': 17.0, 'item_conteudo_gordura': '...   
4  {'item_peso': 8.195, 'item_conteudo_gordura': ...   

                                                loja  
0  {'loja_ano_estabelecimento': 2019, 'loja_taman...  
1  {'loja_ano_estabelecimento': 2009, 'loja_taman...  
2  {'loja_ano_estabelecimento': 2019, 'loja_taman...  
3  {'loja_ano_estabelecimento': 2012, 'loja_taman...  
4  {'loja_ano_estabelecimento': 2019, 'loja_taman...

### Prompt:

O DataFrame "df" possui as colunas "item_identificador", "loja_identificador", "vendas_totais", "item" e "loja".

As colunas "item" e "loja" contêm dicionários aninhados dentro de cada uma das linhas. Como posso transformar os dados dos dicionários e colunas para torná-los mais acessíveis no meu DataFrame?

In [6]:
from pandas import json_normalize

# Achatar a coluna "item"
item_df = json_normalize(df['item'])
item_df.columns = [f"item_{col}" for col in item_df.columns]  # Prefixar colunas para evitar conflitos

# Achatar a coluna "loja"
loja_df = json_normalize(df['loja'])
loja_df.columns = [f"loja_{col}" for col in loja_df.columns]  # Prefixar colunas para evitar conflitos

# Concatenar os DataFrames achatados de volta ao DataFrame original
df = pd.concat([df.drop(['item', 'loja'], axis=1), item_df, loja_df], axis=1)

print("\nDataFrame Transformado:")
df



DataFrame Transformado:


item_identificador loja_identificador  vendas_totais  item_item_peso  \
0                 FDB08             OUT018      176503.58           6.055   
1                 DRQ35             OUT049      185758.20           9.300   
2                 FDD14             OUT018      165983.94          20.700   
3                 FDY37             OUT045      314923.40          17.000   
4                 FDY59             OUT018       64782.34           8.195   
...                 ...                ...            ...             ...   
8545              FDY08             OUT010       28096.76           9.395   
8546              FDC41             OUT017      130163.90          15.600   
8547              NCQ53             OUT045      614533.40          17.600   
8548              FDL46             OUT017      164985.24          20.350   
8549              NCN30             OUT046       96541.00          16.350   

     item_item_conteudo_gordura  item_item_visibilidade        item_item_tipo  \
0         Baixo Teor de Gordura                0.031230     Frutas e Vegetais   
1         Baixo Teor de Gordura                0.042357    Bebidas Alcoólicas   
2         Baixo Teor de Gordura                0.170500             Enlatados   
3                       Regular                0.026623             Enlatados   
4         Baixo Teor de Gordura                0.000000           Confeitaria   
...                         ...                     ...                   ...   
8545                    Regular                0.286345     Frutas e Vegetais   
8546      Baixo Teor de Gordura                0.117575  Alimentos Congelados   
8547      Baixo Teor de Gordura                0.018944             Mercearia   
8548      baixo teor de gordura                0.054363               Lanches   
8549                        BTG                0.016993               Cereais   

      item_item_preco item_item_quantidade_venda  \
0              160.36                       None   
1              123.24                       None   
2              184.13                       None   
3              144.25                       None   
4               93.15                       None   
...               ...                        ...   
8545           139.18                       None   
8546            75.67                       None   
8547           237.36                       None   
8548           117.95                       None   
8549            95.74                       None   

      loja_loja_ano_estabelecimento loja_loja_tamanho  \
0                              2019             Médio   
1                              2009             Médio   
2                              2019             Médio   
3                              2012              None   
4                              2019             Médio   
...                             ...               ...   
8545                           2008              None   
8546                           2017              None   
8547                           2012              None   
8548                           2017              None   
8549                           2007           Pequeno   

     loja_loja_tipo_localizacao       loja_loja_tipo  
0                       Nível 3  Supermercado Tipo 2  
1                       Nível 1  Supermercado Tipo 1  
2                       Nível 3  Supermercado Tipo 2  
3                       Nível 2  Supermercado Tipo 1  
4                       Nível 3  Supermercado Tipo 2  
...                         ...                  ...  
8545                    Nível 3              Mercado  
8546                    Nível 2  Supermercado Tipo 1  
8547                    Nível 2  Supermercado Tipo 1  
8548                    Nível 2  Supermercado Tipo 1  
8549                    Nível 1  Supermercado Tipo 1  

[8550 rows x 13 columns]

# Aula 2 - Exploração dos dados

### Prompt:

O DataFrame "df" contém dados referentes às vendas de diversos produtos em várias lojas do supermercado. São fornecidos IDs de produtos e IDs de lojas que comercializaram esses produtos, o total arrecadado em vendas, bem como descrições sobre o produto vendido e a loja que o vendeu.

Estamos iniciando a etapa de exploração de dados, na qual desejamos conhecer os valores informados e identificar se existem valores ausentes, duplicados e outras inconsistências.

Por favor, construa um código para executar a etapa de exploração de dados usando o Python.

Não é necessário utilizar "print()" para mostrar o resultado das consultas.

In [7]:
# Estatísticas descritivas para colunas numéricas
descriptive_stats = df.describe()

# Estatísticas descritivas para colunas não numéricas
descriptive_stats_non_numeric = df.describe(include=[object])


In [8]:
descriptive_stats

vendas_totais  item_item_peso  item_item_visibilidade  item_item_preco  \
count   8.550000e+03     7081.000000             8550.000000      8550.000000   
mean    2.181949e+05       12.855023                0.066150       141.007453   
std     1.708098e+05        4.643508                0.051578        62.333062   
min     3.329000e+03        4.555000                0.000000        31.290000   
25%     8.349132e+04        8.770000                0.027024        93.787500   
50%     1.794331e+05       12.600000                0.053978       142.935000   
75%     3.100963e+05       16.850000                0.094646       185.760000   
max     1.308696e+06       21.350000                0.328391       266.890000   

       loja_loja_ano_estabelecimento  
count                    8550.000000  
mean                     2007.830409  
std                         8.372541  
min                      1995.000000  
25%                      1997.000000  
50%                      2009.000000  
75%                      2014.000000  
max                      2019.000000

In [9]:
descriptive_stats_non_numeric

item_identificador loja_identificador item_item_conteudo_gordura  \
count                8550               8550                       8550   
unique               1559                 10                          5   
top                 FDG33             OUT027      Baixo Teor de Gordura   
freq                   10                940                       5101   

           item_item_tipo item_item_quantidade_venda loja_loja_tamanho  \
count                8550                          0              6133   
unique                 16                          0                 3   
top     Frutas e Vegetais                        NaN             Médio   
freq                 1234                        NaN              2805   

       loja_loja_tipo_localizacao       loja_loja_tipo  
count                        8550                 8550  
unique                          3                    4  
top                       Nível 3  Supermercado Tipo 1  
freq                         3362                 5593

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8550 entries, 0 to 8549
Data columns (total 13 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   item_identificador             8550 non-null   object 
 1   loja_identificador             8550 non-null   object 
 2   vendas_totais                  8550 non-null   float64
 3   item_item_peso                 7081 non-null   float64
 4   item_item_conteudo_gordura     8550 non-null   object 
 5   item_item_visibilidade         8550 non-null   float64
 6   item_item_tipo                 8550 non-null   object 
 7   item_item_preco                8550 non-null   float64
 8   item_item_quantidade_venda     0 non-null      object 
 9   loja_loja_ano_estabelecimento  8550 non-null   int64  
 10  loja_loja_tamanho              6133 non-null   object 
 11  loja_loja_tipo_localizacao     8550 non-null   object 
 12  loja_loja_tipo                 8550 non-null   o

In [11]:
df.nunique()

item_identificador               1559
loja_identificador                 10
vendas_totais                    3493
item_item_peso                    415
item_item_conteudo_gordura          5
item_item_visibilidade           7880
item_item_tipo                     16
item_item_preco                  4839
item_item_quantidade_venda          0
loja_loja_ano_estabelecimento       9
loja_loja_tamanho                   3
loja_loja_tipo_localizacao          3
loja_loja_tipo                      4
dtype: int64

In [12]:
# Identificar linhas duplicadas
df.duplicated()

# Contar o número de linhas duplicadas
df.duplicated().sum()


27

In [13]:
# Contar valores ausentes em cada coluna
df.isnull().sum()

# Percentual de valores ausentes
(df.isnull().sum() / len(df)) * 100

item_identificador                 0.000000
loja_identificador                 0.000000
vendas_totais                      0.000000
item_item_peso                    17.181287
item_item_conteudo_gordura         0.000000
item_item_visibilidade             0.000000
item_item_tipo                     0.000000
item_item_preco                    0.000000
item_item_quantidade_venda       100.000000
loja_loja_ano_estabelecimento      0.000000
loja_loja_tamanho                 28.269006
loja_loja_tipo_localizacao         0.000000
loja_loja_tipo                     0.000000
dtype: float64

In [18]:
import pandas as pd
import numpy as np

# Supondo que o DataFrame "df" já tenha sido carregado

# Estatísticas descritivas
descriptive_stats = df.describe()
descriptive_stats_non_numeric = df.describe(include=[object])

# Verificação de valores ausentes
missing_values = df.isnull().sum()
missing_values_percentage = (df.isnull().sum() / len(df)) * 100

# Identificação de duplicados
duplicate_rows = df.duplicated()
num_duplicate_rows = df.duplicated().sum()

# Verificação de valores únicos
unique_values = df.nunique()

# Distribuição dos dados
sales_distribution = df['vendas_totais'].describe()
item_category_distribution = df['item_item_tipo'].value_counts()
store_location_distribution = df['loja_loja_tipo_localizacao'].value_counts()

# Identificação de outliers usando IQR para "vendas_totais"
Q1 = df['vendas_totais'].quantile(0.25)
Q3 = df['vendas_totais'].quantile(0.75)
IQR = Q3 - Q1
outliers = df[(df['vendas_totais'] < (Q1 - 1.5 * IQR)) | (df['vendas_totais'] > (Q3 + 1.5 * IQR))]

# Para visualização adicional, você pode armazenar os resultados em um dicionário ou exportar para um arquivo
results = {
    "descriptive_stats": descriptive_stats,
    "descriptive_stats_non_numeric": descriptive_stats_non_numeric,
    "missing_values": missing_values,
    "missing_values_percentage": missing_values_percentage,
    "num_duplicate_rows": num_duplicate_rows,
    "unique_values": unique_values,
    "sales_distribution": sales_distribution,
    "item_category_distribution": item_category_distribution,
    "store_location_distribution": store_location_distribution,
    "outliers": outliers
}
results
# Exemplo de como salvar os resultados em um arquivo CSV
# descriptive_stats.to_csv('descriptive_stats.csv')
# descriptive_stats_non_numeric.to_csv('descriptive_stats_non_numeric.csv')
# pd.DataFrame(missing_values).to_csv('missing_values.csv')
# pd.DataFrame(unique_values).to_csv('unique_values.csv')
# sales_distribution.to_csv('sales_distribution.csv')
# item_category_distribution.to_csv('item_category_distribution.csv')
# store_location_distribution.to_csv('store_location_distribution.csv')
# outliers.to_csv('outliers.csv')


{'descriptive_stats':        vendas_totais  item_item_peso  item_item_visibilidade  item_item_preco  \
 count   8.550000e+03     7081.000000             8550.000000      8550.000000   
 mean    2.181949e+05       12.855023                0.066150       141.007453   
 std     1.708098e+05        4.643508                0.051578        62.333062   
 min     3.329000e+03        4.555000                0.000000        31.290000   
 25%     8.349132e+04        8.770000                0.027024        93.787500   
 50%     1.794331e+05       12.600000                0.053978       142.935000   
 75%     3.100963e+05       16.850000                0.094646       185.760000   
 max     1.308696e+06       21.350000                0.328391       266.890000   
 
        loja_loja_ano_estabelecimento  
 count                    8550.000000  
 mean                     2007.830409  
 std                         8.372541  
 min                      1995.000000  
 25%                      1997.000000  

In [19]:
df['item_item_conteudo_gordura'].unique()

array(['Baixo Teor de Gordura', 'Regular', 'BTG', 'reg',
       'baixo teor de gordura'], dtype=object)

### Prompt:

O DataFrame "df" contém dados relacionados às vendas durante um trimestre de um supermercado. Desejo utilizá-lo para verificar o desempenho dessas vendas.

Durante a exploração do "df", identificamos algumas inconsistências, como:

- Linhas duplicadas,
- Valores nulos e
- Dados categóricos que foram escritos de maneira não padronizada, dando a ideia de que existem mais categorias do que as que realmente existem.

O que devo fazer para garantir que essas inconsistências que encontrei não afetem minha análise?

# Aula 3 - Limpeza e tratamento

### Prompt:

Durante a exploração do DataFrame "df", identificamos 27 linhas duplicadas usando "df.duplicated().sum()".

O objetivo é analisar os dados em "df" para medir o desempenho de uma empresa durante um trimestre de vendas.

O que devemos fazer com os dados duplicados para garantir que não haja impacto negativo na análise?

### Prompt:

Identificamos três colunas em "df" que apresentam valores nulos:

- A coluna "item_quantidade_venda" possui 8550 linhas com valores nulos.
- A coluna "item_peso" possui 1469 linhas com valores nulos.
- A coluna "loja_tamanho" possui 2417 linhas com valores nulos.

Houve um grande número de vendas de itens durante o trimestre, e não há justificativa para a ausência de dados na coluna "item_quantidade_venda", o que a torna uma coluna dispensável.

A coluna "item_peso" contém valores do tipo float, representando o peso em gramas dos produtos.

A coluna "loja_tamanho" possui três categorias de tamanho para supermercados.

Como devo lidar com esses dados nulos para garantir que minha análise não seja prejudicada?

### Prompt:

A coluna "item_conteudo_gordura" do DataFrame "df" apresenta 5 valores únicos dados por: 'Baixo Teor de Gordura', 'Regular', 'BTG', 'reg' e 'baixo teor de gordura'.

Os valores únicos da coluna "item_conteudo_gordura" deveriam ser apenas 'Baixo Teor de Gordura' e 'Regular', mas foram inseridos sem padronização, o que resultou em mais categorias do que o desejado.

Como posso ajustar os dados da coluna "item_conteudo_gordura" para manter apenas as duas categorias padrões?

# Aula 4 - Visualização dos dados

### Prompt:

Os dados do DataFrame "df" estão limpos e tratados. Agora, desejo iniciar a etapa de construção de gráficos para criar visualizações que ofereçam um bom suporte à análise do desempenho de vendas da empresa.

As colunas disponíveis em "df" são as seguintes:
- "item_identificador": Número único atribuído a cada Item
- "loja_identificador": Nome único da Loja
- "vendas_totais": Valor obtido em vendas totais
- "item_peso": Peso do Item em gramas
- "item_conteudo_gordura": Conteúdo de Gordura do Item
- "item_visibilidade": Valor de posicionamento de cada item.
- "item_tipo": Tipo de utilidade do item
- "item_preço": Preço unitário do item
- "loja_ano_estabelecimento": Ano de Estabelecimento da Loja.
- "loja_tamanho": Tamanho da Loja
- "loja_tipo_localização": Tipo de localização da loja
- "loja_tipo": Tipo de loja.

O objetivo é encontrar padrões ou informações que se destacam quando nos referimos ao total obtido em vendas especificado na coluna "vendas_totais".

Com base nas informações fornecidas, quais tipos de visualizações seriam úteis para atingir o objetivo de analisar o desempenho de vendas?

### Prompt:

O DataFrame "df" contém a coluna "vendas_totais", que informa o valor em reais obtido em vendas de produtos por várias lojas.

Além disso, em "df", temos a coluna "loja_tamanho" com 4 categorias de tamanho para lojas: 'Grande', 'Médio', 'Pequeno' e 'Não registrado'.

Desejo criar uma visualização que permita identificar o total obtido em vendas para as quatro categorias de tamanho de loja e representá-las em um gráfico de barras.

Os dados na coluna "vendas_totais" precisam ser agrupados pelas categorias de "loja_tamanho".

A figura da plotagem deve seguir a identidade visual da loja, com o fundo da figura e do plot na cor "#ebebeb" e as barras na cor "#e23155".

Além disso, é necessário exibir o total das vendas em milhões acima de cada coluna, com o texto dessa informação na cor "#19325c" e a estrutura "R$ valor M".

Utilize as bibliotecas de visualização de dados do Python para criar essa visualização.

### Prompt:

O DataFrame "df" contém a coluna "vendas_totais", que informa o valor em reais obtido em vendas de produtos por várias lojas.

Além disso, em "df", temos a coluna "loja_tipo" com 4 categorias de tipo para lojas: 'Supermercado Tipo 1', 'Supermercado Tipo 2',  'Mercado', 'Supermercado Tipo 3'.

Desejo criar uma visualização que permita identificar o total obtido em vendas para as quatro categorias de tipo de loja e representá-las em um gráfico de barras.

Os dados na coluna "vendas_totais" precisam ser agrupados pelas categorias de "loja_tipo".

A figura da plotagem deve seguir a identidade visual da loja, com o fundo da figura e do plot na cor "#ebebeb" e as barras na cor "#e23155".

Além disso, é necessário exibir o total das vendas em milhões acima de cada coluna, com o texto dessa informação na cor "#19325c" e a estrutura "R$ valor M".

Utilize as bibliotecas de visualização de dados do Python para criar essa visualização.

### Prompt:

O DataFrame "df" contém a coluna "vendas_totais", que informa o valor em reais obtido em vendas de produtos por várias lojas.

Além disso, em "df", temos a coluna "item_tipo" com 16 categorias de produtos.

Desejo criar uma visualização que permita identificar o total obtido em vendas por categorias de produtos em um gráfico de barras horizontais.

Os dados na coluna "vendas_totais" precisam ser agrupados pelas categorias de "item_tipo".

A figura da plotagem deve seguir a identidade visual da loja, com o fundo da figura e do plot na cor "#ebebeb".

As barras precisam estar na cor "#e23155".

Os valores das barras precisam estar ordenadas para manter um aspecto de funil.

Além disso, é necessário exibir o total das vendas em milhões dentro ou ao lado de cada barra, com o texto dessa informação na cor "#19325c" e na estrutura "R$ valor M".

Não mostre o grid no plot.

Utilize as bibliotecas de visualização de dados do Python para criar essa visualização.

### Prompt:

O DataFrame "df" contém a coluna "vendas_totais", que informa o valor em reais obtido em vendas de produtos por várias lojas.

Além disso, em "df", temos a coluna "item_conteudo_gordura" com 2 categorias: 'Baixo Teor de Gordura' e 'Regular'.

Desejo criar uma visualização que permita identificar o total obtido em vendas para as 2 categorias de quantidade de gordura em um gráfico de coluna.

Os dados na coluna "vendas_totais" precisam ser agrupados pelas categorias de "item_conteudo_gordura".

A figura da plotagem deve seguir a identidade visual da loja, com o fundo da figura e do plot na cor "#ebebeb".

As colunas de cada categoria devem ter cores específicas: a categoria 'Baixo Teor de Gordura' na cor "#e23155" e a categoria 'Regular' na cor "#acdcd3".

Além disso, é necessário exibir o total das vendas em milhões acima de cada coluna, com o texto dessa informação na cor "#19325c" e na estrutura "R$ valor M".

Utilize as bibliotecas de visualização de dados do Python para criar essa visualização.